File: 7.API-DESIGN-RestAPI/code/typescript/filtering_3.ts
import { Request, Response } from 'express';

// GET /v1/organizations?status=active&sortBy=name&sortOrder=ascending&page=1&limit=10
export async function listOrganizations(req: Request, res: Response) {
    // --- sane defaults: never crash if the client omits params ---
    const page = parseInt(req.query.page as string) || 1;
    const limit = parseInt(req.query.limit as string) || 10;
    const sortBy = (req.query.sortBy as string) || 'createdAt';
    const sortOrder = (req.query.sortOrder as string) || 'descending';

    const filters: Record<string, string> = {};
    if (req.query.status) {
        filters['status'] = req.query.status as string; // ?status=active
    }

    const { rows, total } = await store.query(filters, sortBy, sortOrder, page, limit);
    const totalPages = Math.ceil(total / limit); // ceil division

    return res.status(200).json({
        data: rows,
        total: total,
        page: page,
        totalPages: totalPages
    });
}

// Mock store for compilation
const store = {
    query: async (filters: any, sortBy: string, sortOrder: string, page: number, limit: number): Promise<any> => ({ rows: [], total: 0 })
};

File: 7.API-DESIGN-RestAPI/code/rust/filtering_3.rs
use actix_web::{web, HttpResponse, Responder};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct ListQuery {
    pub status: Option<String>,
    pub sort_by: Option<String>,
    pub sort_order: Option<String>,
    pub page: Option<u32>,
    pub limit: Option<u32>,
}

// GET /v1/organizations?status=active&sortBy=name&sortOrder=ascending&page=1&limit=10
pub async fn list_organizations(query: web::Query<ListQuery>) -> impl Responder {
    // --- sane defaults: never crash if the client omits params ---
    let page = query.page.unwrap_or(1);
    let limit = query.limit.unwrap_or(10);
    let sort_by = query.sort_by.clone().unwrap_or_else(|| "createdAt".to_string());
    let sort_order = query.sort_order.clone().unwrap_or_else(|| "descending".to_string());

    let mut filters = std::collections::HashMap::new();
    if let Some(ref s) = query.status {
        filters.insert("status".to_string(), s.clone()); // ?status=active
    }

    let (rows, total) = store_query(filters, &sort_by, &sort_order, page, limit).await;
    let total_pages = (total + limit - 1) / limit; // ceil division

    HttpResponse::Ok().json(serde_json::json!({
        "data": rows,
        "total": total,
        "page": page,
        "totalPages": total_pages,
    }))
}

// Mock
async fn store_query(_filters: std::collections::HashMap<String, String>, _sort_by: &str, _sort_order: &str, _page: u32, _limit: u32) -> (Vec<serde_json::Value>, u32) {
    (vec![], 0)
}
